core/util/src/main/java/org/eclipse/rdf4j/common/text/StringUtil.java

```diff
@@ -72,8 +82,61 @@ public static String gsub(String olds, String news, String text) {
 		return buf.toString();
 	}
 
+	private static String hex(int c) {
+		return Integer.toHexString(c).toUpperCase(Locale.US);
+	}
+
+	/**
+	 * Escapes string to a (mostly) conforming IRI value and append it to the appendable. Most invalid characters are
+	 * percent-encoded, non-ASCII values can optionally be numerically encoded.
```

Comment:
See also https://www.w3.org/TR/turtle/#numeric

Reply:
Given that both Turtle and N-Triples specs now specify UTF-8 rather than ASCII, and IRIs supercede URIs (see RFC 3987 vs 3986), I think we should, when serializing, consistently use numeric escapes. %-encodings should just not be touched.

Reply:
I was probably not clear enough in the javadoc :-)

The code does exactly what you suggested: using numeric escapes (IF escapeEncoding is set, which can be the case for NTriples since it still allows for ASCII served over legacy text/plain. Turtle is all UTF8 so it doesn't use this setting)

For invalid characters in IRIs (thus an invalid IRI to begin with...), like a space or a weird control character, % encoding is used (otherwise there is no way to make it parse, numeric encoding will still fail)

Reply:
Ah I see! Right that makes more sense, thanks for clarifying :)